### DISCUSSION TITLE
PnG to JP2

### DISCUSSION THREAD

**Discussion by khawar-islam** [ImageMagick/ImageMagick#2747] [Help]
I am keep trying to convert png images into jpe2000 format. The main problem is quality factor which is not working

`    convert test/images/kodim$i.png -quality $(($j*5)) test/JP2/kodim$i/`printf "%02d" $j`.jp2
`
Complete code

```
for i in {01..24..1}; do
  echo JP2 Encoding test/images/kodim$i.png
  mkdir -p test/JP2/kodim$i
  for j in {0..20..1}; do
    convert test/images/kodim$i.png -quality $(($j*5)) test/JP2/kodim$i/`printf "%02d" $j`.jp2
  done
done
```

**Comment by snibgo** [comment]
What version of IM, on what platform?

If you use IM v7, I suggest you use "magick", not "convert".

> The main problem is quality factor which is not working

What goes wrong? How do you know it is not working?

"-quality" works fine for me, with IM v7.0.8-64. From an ordinary photo, different quality numbers give files of different sizes, and visibly different quality at low "-quality" settings.

**Reply by khawar-islam** [reply]
The version: ImageMagick 6.9.7-4 Q16 x86_64 20170114 http://www.imagemagick.org

I will tell in detail
`convert test/images/kodim$i.png -quality $(($j*5)) -sampling-factor 4:2:0 test/jpeg/kodim$i/`printf "%02d" $j`.jpg`

When i entered above command, it produces 20 pictures with different qualities like starting from picture 1 no clear then picture 2 is more clearer than picture and this process go so on and image 20 is more clear and near to test image dataset. this above command works fine.
Now, I want to do same for JP2 with quality factor. When i entered JP2 command, it produces same images with high resolution even in first image.Not similar with like starting from low resolution and goes to high resolution. I also share jpeg and JP2 images. If you observe the quality of JPEG images it get better and better in one image. It fails on JP2.
[kodim01.zip](https://github.com/ImageMagick/ImageMagick/files/5409474/kodim01.zip)
[kodim01JP2.zip](https://github.com/ImageMagick/ImageMagick/files/5409480/kodim01JP2.zip)

**Reply by khawar-islam** [reply]
Even i installed new updated version of imagemagick

**Comment by snibgo** [comment]
What does "convert -version" say? And "convert -format"?

In kodim02JP2.zip, you have files like "00.jp2" but they are all PNG format, not JP2.

I suspect that your build of IM does not write to JP2, so the format is unchanged from the input.

**Reply by khawar-islam** [reply]
One folder is PNG and another one Jp2 format. Let suppose, we have 1 pic and imagemagick generates 20 picture with different qualities. Now version is Version: ImageMagick 7.0.8-11 Q16

**Comment by fmw42** [comment]
Do you have OpenJPEG installed and if so, what version? 

`magick -version`

should contain jp2

magick -list format 

should have a line such as

    ` JP2* JP2       rw-   JPEG-2000 File Format Syntax (2.3.1)`

The version number is at the end.

I believe you need at least version 2.1.0

See https://imagemagick.org/script/jp2.php for the proper format for setting the quality.

**Reply by khawar-islam** [reply]
magick -version
Version: ImageMagick 7.0.8-11 Q16 x86_64 2018-08-31 https://www.imagemagick.org
Copyright: © 1999-2018 ImageMagick Studio LLC
License: https://www.imagemagick.org/script/license.php
Features: Cipher DPC HDRI 
Delegates (built-in): bzlib fftw fontconfig freetype gvc jbig jng jp2 jpeg lzma pangocairo png tiff webp x xml zlib


(RNN_Fresh) khawar@khawar-B365-HD3:~/Desktop/RNN_Fresh/pytorch-image-comp-rnn$ magick -list format
   Format  Mode  Description
-------------------------------------------------------------------------------
      3FR  r--   Hasselblad CFV/H3D39II
      3G2  r--   Media Container
      3GP  r--   Media Container
        A* rw+   Raw alpha samples
      AAI* rw+   AAI Dune image
       AI  rw-   Adobe Illustrator CS2
      ART* rw-   PFS: 1st Publisher Clip Art
      ARW  r--   Sony Alpha Raw Image Format
      AVI  r--   Microsoft Audio/Visual Interleaved
      AVS* rw+   AVS X image
        B* rw+   Raw blue samples
      BGR* rw+   Raw blue, green, and red samples
     BGRA* rw+   Raw blue, green, red, and alpha samples
     BGRO* rw+   Raw blue, green, red, and opacity samples
      BIE* rw-   Joint Bi-level Image experts Group interchange format (2.1)
      BMP* rw-   Microsoft Windows bitmap image
     BMP2* rw-   Microsoft Windows bitmap image (V2)
     BMP3* rw-   Microsoft Windows bitmap image (V3)
      BRF* -w-   BRF ASCII Braille format
        C* rw+   Raw cyan samples
      CAL* rw-   Continuous Acquisition and Life-cycle Support Type 1
           Specified in MIL-R-28002 and MIL-PRF-28002
     CALS* rw-   Continuous Acquisition and Life-cycle Support Type 1
           Specified in MIL-R-28002 and MIL-PRF-28002
   CANVAS* r--   Constant image uniform color
  CAPTION* r--   Caption
      CIN* rw-   Cineon Image File
      CIP* -w-   Cisco IP phone image format
     CLIP* rw+   Image Clip Mask
     CMYK* rw+   Raw cyan, magenta, yellow, and black samples
    CMYKA* rw+   Raw cyan, magenta, yellow, black, and alpha samples
      CR2  r--   Canon Digital Camera Raw Image Format
      CRW  r--   Canon Digital Camera Raw Image Format
      CUR* rw-   Microsoft icon
      CUT* r--   DR Halo
      DCM* r--   Digital Imaging and Communications in Medicine image
           DICOM is used by the medical community for images like X-rays.  The
           specification, "Digital Imaging and Communications in Medicine
           (DICOM)", is available at http://medical.nema.org/.  In particular,
           see part 5 which describes the image encoding (RLE, JPEG, JPEG-LS),
           and supplement 61 which adds JPEG-2000 encoding.
      DCR  r--   Kodak Digital Camera Raw Image File
    DCRAW  r--   Raw Photo Decoder (dcraw)
      DCX* rw+   ZSoft IBM PC multi-page Paintbrush
      DDS* rw+   Microsoft DirectDraw Surface
    DFONT* r--   Multi-face font package (Freetype 2.9.1)
      DNG  r--   Digital Negative
      DPX* rw-   SMPTE 268M-2003 (DPX 2.0)
           Digital Moving Picture Exchange Bitmap, Version 2.0.
           See SMPTE 268M-2003 specification at http://www.smtpe.org
           
     DXT1* rw+   Microsoft DirectDraw Surface
     DXT5* rw+   Microsoft DirectDraw Surface
     EPDF  rw-   Encapsulated Portable Document Format
      EPI  rw-   Encapsulated PostScript Interchange format
      EPS  rw-   Encapsulated PostScript
     EPS2  -w-   Level II Encapsulated PostScript
     EPS3  -w+   Level III Encapsulated PostScript
     EPSF  rw-   Encapsulated PostScript
     EPSI  rw-   Encapsulated PostScript Interchange format
      EPT  rw-   Encapsulated PostScript with TIFF preview
     EPT2  rw-   Encapsulated PostScript Level II with TIFF preview
     EPT3  rw+   Encapsulated PostScript Level III with TIFF preview
      ERF  r--   Epson RAW Format
      FAX* rw+   Group 3 FAX
           FAX machines use non-square pixels which are 1.5 times wider than
           they are tall but computer displays use square pixels, therefore
           FAX images may appear to be narrow unless they are explicitly
           resized using a geometry of "150x100%".
           
     FILE* r--   Uniform Resource Locator (file://)
     FITS* rw-   Flexible Image Transport System
  FRACTAL* r--   Plasma fractal image
      FTP* r--   Uniform Resource Locator (ftp://)
      FTS* rw-   Flexible Image Transport System
        G* rw+   Raw green samples
       G3* rw-   Group 3 FAX
       G4* rw-   Group 4 FAX
      GIF* rw+   CompuServe graphics interchange format
    GIF87* rw-   CompuServe graphics interchange format (version 87a)
 GRADIENT* r--   Gradual linear passing from one shade to another
     GRAY* rw+   Raw gray samples
    GRAYA* rw+   Raw gray and alpha samples
   GROUP4* rw-   Raw CCITT Group4
     HALD* r--   Identity Hald color lookup table image
      HDR* rw+   Radiance RGBE image format
HISTOGRAM* -w-   Histogram of the image
      HRZ* rw-   Slow Scan TeleVision
      HTM* -w-   Hypertext Markup Language and a client-side image map
     HTML* -w-   Hypertext Markup Language and a client-side image map
     HTTP* r--   Uniform Resource Locator (http://)
    HTTPS* ---   Uniform Resource Locator (https://)
      ICB* rw-   Truevision Targa image
      ICO* rw+   Microsoft icon
     ICON* rw-   Microsoft icon
      IIQ  r--   Phase One Raw Image Format
     INFO  -w+   The image format and characteristics
   INLINE* rw+   Base64-encoded inline images
      IPL* rw+   IPL Image Sequence
   ISOBRL* -w-   ISO/TR 11548-1 format
  ISOBRL6* -w-   ISO/TR 11548-1 format 6dot
      J2C* rw-   JPEG-2000 Code Stream Syntax (2.3.1)
      J2K* rw-   JPEG-2000 Code Stream Syntax (2.3.1)
      JBG* rw+   Joint Bi-level Image experts Group interchange format (2.1)
     JBIG* rw+   Joint Bi-level Image experts Group interchange format (2.1)
      JNG* rw-   JPEG Network Graphics
           See http://www.libpng.org/pub/mng/ for details about the JNG
           format.
      JNX* r--   Garmin tile format
      JP2* rw-   JPEG-2000 File Format Syntax (2.3.1)
      JPC* rw-   JPEG-2000 Code Stream Syntax (2.3.1)
      JPE* rw-   Joint Photographic Experts Group JFIF format (90)
     JPEG* rw-   Joint Photographic Experts Group JFIF format (90)
      JPG* rw-   Joint Photographic Experts Group JFIF format (90)
      JPM* rw-   JPEG-2000 File Format Syntax (2.3.1)
      JPS* rw-   Joint Photographic Experts Group JFIF format (90)
      JPT* rw-   JPEG-2000 File Format Syntax (2.3.1)
     JSON  -w+   The image format and characteristics
        K* rw+   Raw black samples
      K25  r--   Kodak Digital Camera Raw Image Format
      KDC  r--   Kodak Digital Camera Raw Image Format
    LABEL* r--   Image label
        M* rw+   Raw magenta samples
      M2V  rw+   MPEG Video Stream
      M4V  rw+   Raw MPEG-4 Video
      MAC* r--   MAC Paint
      MAP* rw-   Colormap intensities and indices
     MASK* rw+   Image Clip Mask
      MAT  rw+   MATLAB level 5 image format
    MATTE* -w+   MATTE format
      MEF  r--   Mamiya Raw Image File
     MIFF* rw+   Magick Image File Format
      MKV  rw+   Multimedia Container
      MNG* rw+   Multiple-image Network Graphics (libpng 1.6.34,1.6.37)
           See http://www.libpng.org/pub/mng/ for details about the MNG
           format.
     MONO* rw-   Raw bi-level bitmap
      MOV  rw+   MPEG Video Stream
      MP4  rw+   MPEG-4 Video Stream
      MPC* rw+   Magick Persistent Cache image format
     MPEG  rw+   MPEG Video Stream
      MPG  rw+   MPEG Video Stream
      MRW  r--   Sony (Minolta) Raw Image File
      MSL* rw+   Magick Scripting Language
     MSVG  rw+   ImageMagick's own SVG internal renderer
      MTV* rw+   MTV Raytracing image format
      MVG* rw-   Magick Vector Graphics
      NEF  r--   Nikon Digital SLR Camera Raw Image File
      NRW  r--   Nikon Digital SLR Camera Raw Image File
     NULL* rw-   Constant image of uniform color
        O* rw+   Raw opacity samples
      ORF  r--   Olympus Digital Camera Raw Image File
      OTB* rw-   On-the-air bitmap
      OTF* r--   Open Type font (Freetype 2.9.1)
      PAL* rw-   16bit/pixel interleaved YUV
     PALM* rw+   Palm pixmap
      PAM* rw+   Common 2-dimensional bitmap format
    PANGO* r--   Pango Markup Language (Pangocairo 1.40.14)
  PATTERN* r--   Predefined pattern
      PBM* rw+   Portable bitmap format (black and white)
      PCD* rw-   Photo CD
     PCDS* rw-   Photo CD
      PCL  rw+   Printer Control Language
      PCT* rw-   Apple Macintosh QuickDraw/PICT
      PCX* rw-   ZSoft IBM PC Paintbrush
      PDB* rw+   Palm Database ImageViewer Format
      PDF  rw+   Portable Document Format
     PDFA  rw+   Portable Document Archive Format
      PEF  r--   Pentax Electronic File
      PES* r--   Embrid Embroidery Format
      PFA* r--   Postscript Type 1 font (ASCII) (Freetype 2.9.1)
      PFB* r--   Postscript Type 1 font (binary) (Freetype 2.9.1)
      PFM* rw+   Portable float format
      PGM* rw+   Portable graymap format (gray scale)
      PGX* rw-   JPEG 2000 uncompressed format
    PICON* rw-   Personal Icon
     PICT* rw-   Apple Macintosh QuickDraw/PICT
      PIX* r--   Alias/Wavefront RLE image format
    PJPEG* rw-   Joint Photographic Experts Group JFIF format (90)
   PLASMA* r--   Plasma fractal image
      PNG* rw-   Portable Network Graphics (libpng 1.6.34,1.6.37)
           See http://www.libpng.org/ for details about the PNG format.
    PNG00* rw-   PNG inheriting bit-depth, color-type from original, if possible
    PNG24* rw-   opaque or binary transparent 24-bit RGB (zlib 1.2.11)
    PNG32* rw-   opaque or transparent 32-bit RGBA
    PNG48* rw-   opaque or binary transparent 48-bit RGB
    PNG64* rw-   opaque or transparent 64-bit RGBA
     PNG8* rw-   8-bit indexed with optional binary transparency
      PNM* rw+   Portable anymap
      PPM* rw+   Portable pixmap format (color)
       PS  rw+   PostScript
      PS2  -w+   Level II PostScript
      PS3  -w+   Level III PostScript
      PSB* rw+   Adobe Large Document Format
      PSD* rw+   Adobe Photoshop bitmap
     PTIF* rw+   Pyramid encoded TIFF
      PWP* r--   Seattle Film Works
        R* rw+   Raw red samples
RADIAL-GRADIENT* r--   Gradual radial passing from one shade to another
      RAF  r--   Fuji CCD-RAW Graphic File
      RAS* rw+   SUN Rasterfile
      RAW  r--   Raw
      RGB* rw+   Raw red, green, and blue samples
     RGBA* rw+   Raw red, green, blue, and alpha samples
     RGBO* rw+   Raw red, green, blue, and opacity samples
      RGF* rw-   LEGO Mindstorms EV3 Robot Graphic Format (black and white)
      RLA* r--   Alias/Wavefront image
      RLE* r--   Utah Run length encoded image
      RMF  r--   Raw Media Format
      RW2  r--   Panasonic Lumix Raw Image
      SCR* r--   ZX-Spectrum SCREEN$
SCREENSHOT* r--   Screen shot
      SCT* r--   Scitex HandShake
      SFW* r--   Seattle Film Works
      SGI* rw+   Irix RGB image
    SHTML* -w-   Hypertext Markup Language and a client-side image map
      SIX* rw-   DEC SIXEL Graphics Format
    SIXEL* rw-   DEC SIXEL Graphics Format
SPARSE-COLOR* -w+   Sparse Color
      SR2  r--   Sony Raw Format 2
      SRF  r--   Sony Raw Format
  STEGANO* r--   Steganographic image
      SUN* rw+   SUN Rasterfile
      SVG  rw+   Scalable Vector Graphics (XML 2.9.8)
     SVGZ  rw+   Compressed Scalable Vector Graphics (XML 2.9.8)
     TEXT* r--   Text
      TGA* rw-   Truevision Targa image
THUMBNAIL* -w+   EXIF Profile Thumbnail
     TIFF* rw+   Tagged Image File Format (LIBTIFF, Version 4.1.0)
   TIFF64* rw-   Tagged Image File Format (64-bit) (LIBTIFF, Version 4.1.0)
     TILE* r--   Tile image with a texture
      TIM* r--   PSX TIM
      TTC* r--   TrueType font collection (Freetype 2.9.1)
      TTF* r--   TrueType font (Freetype 2.9.1)
      TXT* rw+   Text
     UBRL* -w-   Unicode Text format
    UBRL6* -w-   Unicode Text format 6dot
      UIL* -w-   X-Motif UIL table
     UYVY* rw-   16bit/pixel interleaved YUV
      VDA* rw-   Truevision Targa image
    VICAR* rw-   VICAR rasterfile format
      VID* rw+   Visual Image Directory
     VIFF* rw+   Khoros Visualization image
     VIPS* rw+   VIPS image
      VST* rw-   Truevision Targa image
     WBMP* rw-   Wireless Bitmap (level 0) image
     WEBP* rw-   WebP Image Format (libwebp 0.5.2 [0208])
      WMV  rw+   Windows Media Video
      WPG* r--   Word Perfect Graphics
        X* rw+   X Image
      X3F  r--   Sigma Camera RAW Picture File
      XBM* rw-   X Windows system bitmap (black and white)
       XC* r--   Constant image uniform color
      XCF* r--   GIMP image
      XPM* rw-   X Windows system pixmap (color)
      XPS  r--   Microsoft XML Paper Specification
       XV* rw+   Khoros Visualization image
      XWD* rw-   X Windows system window dump (color)
        Y* rw+   Raw yellow samples
    YCbCr* rw+   Raw Y, Cb, and Cr samples
   YCbCrA* rw+   Raw Y, Cb, Cr, and alpha samples
      YUV* rw-   CCIR 601 4:1:1 or 4:2:2

* native blob support
r read support
w write support
+ support for multiple images

**Reply by khawar-islam** [reply]
Like this command convert images into jp2 but never loses any information. How is it possible ? There is major difference between PNG and JPEG2000
`magick test/images/kodim$i.png test/JP2/kodim$i/`printf "%02d" $j`.jp2`

**Comment by snibgo** [comment]
> JP2* rw- JPEG-2000 File Format Syntax (2.3.1)

Good.

I am confused. What problem are you trying to solve? Please show examples of the problem, with a single input file, and show that file.

**Comment by fmw42** [comment]
From https://imagemagick.org/script/jp2.php, the quality value is PSNR in db not percent.

**Reply by khawar-islam** [reply]
Simply, i want to convert images with different quality.Simple question

**Reply by t246246** [reply]
It seems -define jp2:quality=value is working.

$ magick logo: -define jp2:quality=30 logo-30.jp2
$ magick logo: -define jp2:quality=50 logo-50.jp2
$ magick compare logo: logo-30.jp2 -metric PSNR -format '%c' NUL:
30.5446
$ magick compare logo: logo-50.jp2 -metric PSNR -format '%c' NUL:
49.4042

**Reply by khawar-islam** [reply]
Thank you. It works

Is it possible that we define sampling factor?

**Reply by t246246** [reply]
I'm not familiar with jpeg 2000, however, this document says

JPEG-2000 does not explicitly define a method of subsampling color components as JPEG  does

http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.93.7046&rep=rep1&type=pdf

**Reply by khawar-islam** [reply]
Overall. I am thankful to your time and guidance.
